File: BFS_and_DFS/programmers43162.cpp
// 문제 및 풀이: https://velog.io/@e7838752/programmers43162
#include <string>
#include <vector>
#include <set>

using namespace std;

int findParent(int x, int parent[]) {
    if (x == parent[x]) return x;
    else return parent[x] = findParent(parent[x], parent);
}

void unionParent(int a, int b, int parent[]) {
    a = findParent(a, parent);
    b = findParent(b, parent);
    
    if (a < b) parent[b] = a;
    else parent[a] = b;
}

int solution(int n, vector<vector<int>> computers) {
    int parent[200];
    for (int i = 0; i < n; ++i) {
        parent[i] = i;
    }
    
    for (int i = 0; i < computers.size(); ++i) {
        for (int j = 0; j < i; ++j) {
            if (computers[i][j] == 1) unionParent(i, j, parent);
        }
    }
    
    set<int> s;
    for (int i = 0; i < n; ++i) {
        s.insert(findParent(i, parent));
    }
    return s.size();
}

File: BFS_and_DFS/programmers43162.js
function solution(n, computers) {
  var answer = 0;
  let parent = Array(n + 1);
  for (let i = 0; i < n; i++) {
    parent[i] = i;
  }

  function findParent(x) {
    if (parent[x] === x) return x;
    return (parent[x] = findParent(parent[x]));
  }

  function unionParent(a, b) {
    a = findParent(a);
    b = findParent(b);

    if (a < b) parent[b] = a;
    else parent[a] = b;
  }

  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n; j++) {
      if (i === j) continue;
      if (computers[i][j] === 1) unionParent(i, j);
    }
  }

  const hs = new Map();
  for (let i = 0; i < n; i++) {
    const x = findParent(i);
    hs.set(x, (hs.get(x) || 0) + 1);
  }

  answer = hs.size;
  return answer;
}
